FixMath2.h

```diff
@@ -0,0 +1,845 @@
+  */
+  SFixMath2(double fl)  {internal_value = static_cast<internal_type> (fl * (next_greater_type(1) << NF)); }
+
+
+    /** Constructor from an integer value which can be interpreted as both a resulting fixed point 
+math number with a fractional part equals to 0, or as a number with decimal, ie as the underlying type behind the fixed point math number.
+@param value Integer value
+@param as_raw=false with false value will be interpreted as an integer, with true it will be interpreted as a number with decimals.
+@return An signed fixed point number
+  */
+  template<typename T>
+  SFixMath2(T value,bool as_raw=false)
+  {
+    // if (as_frac) internal_value = value;
+    if (as_raw) fromRaw(value);
+    else internal_value = (internal_type(value) << NF);
+  }
+
+    /** Set the internal value of the fixed point math number.
+@param raw The new internal value.
+  */  
+  template<typename T>
+  void fromRaw(T raw) { internal_value = raw; }
+
+
+  /** Constructor from another sFixMath2. 
+@param uf A signed fixed type number which value can be represented in this type.
+@return A signed fixed type number
+  */
+  template<byte _NI, byte _NF>
+  SFixMath2(const SFixMath2<_NI,_NF>& uf) {
+    internal_value = SHIFTR((typename IntegerType<((MAX(NI+NF-1,_NI+_NF-1))>>3)+1>::unsigned_type) uf.asRaw(),(_NF-NF));
+  }
+
+  /** Constructor from another UFixMath2. 
+@param uf A unsigned fixed type number.
+@return A signed fixed type number
+  */
+  template<byte _NI, byte _NF>
+  SFixMath2(const UFixMath2<_NI,_NF>& uf) {
+    internal_value = SHIFTR((typename IntegerType<((MAX(NI+NF-1,_NI+_NF-1))>>3)+1>::unsigned_type) uf.asRaw(),(_NF-NF));
+  }
+
+  //////// ADDITION OVERLOADS
+
+    /** Addition with another SFixMath2. Safe.
+@param op The SFixMath2 to be added.
+@return The result of the addition as a SFixMath2.
+  */
+  template<byte _NI, byte _NF>
+  SFixMath2<MAX(NI,_NI)+1,MAX(NF,_NF)> operator+ (const SFixMath2<_NI,_NF>& op) const
+  {
+    constexpr byte new_NI = MAX(NI, _NI) + 1;
+    constexpr byte new_NF = MAX(NF, _NF);
+    typedef typename IntegerType< ((new_NI+new_NF-1)>>3)+1>::unsigned_type return_type;
+    SFixMath2<new_NI,new_NF> left(*this);
+    SFixMath2<new_NI,new_NF> right(op);
+
+    return_type tt = return_type(left.asRaw()) + right.asRaw();
+    return SFixMath2<new_NI,new_NF>(tt,true);
+  }
+
+  /** Addition with another type. Unsafe
+@param op The number to be added.
+@return The result of the addition as a UFixMath2.
+  */
+  template<typename T>
+  SFixMath2<NI,NF> operator+ (const T op) const
+  {
+    return SFixMath2<NI,NF>(internal_value+(op<<NF),true);
+  }
+
+
+  //////// SUBSTRACTION OVERLOADS
+
+       /** Subtraction with another SFixMath2. Safe.
+@param op The SFixMath2 to be subtracted.
+@return The result of the subtraction as a SFixMath2.
+  */ 
+  template<byte _NI, byte _NF> // We do not have the +1 after MAX(NI, _NI) because the substraction between two SFix should fit in the bigger of the two.
```

Comment:
Not necessarily true in the signed case.

Reply:
Wooooooooooops, Very good catch!!

Reply:
Following this, I think all the substraction between UFix and SFix should get a `+2` and not a `+1`? If I get my maths right…

Reply:
Or even more precisely: `SFixMath2<MAX(NI,_NI+1)+1,MAX(NF,_NF)> operator- (const SFixMath2<NI,NF>& op1, const UFixMath2<_NI,_NF>& op2 )`

Reply:
Alternative way of thinking about it: Don't count the sign bit in NI, but only take it into account when determining the required storage type. I.e. a "plain" int16_t would correspond to SFixMath2<15,0>.

I think that should make the math easier?